Rootfs btrfs

jakogut opened the pull request:
This PR implements compressed BTRFS support on top of the rootfs-agnostic branch. It serves as a demonstration of adding support for other filesystems, as well as a test case for benchmarks.

Several commits are WIP, and aren't yet suitable to be merged.

---
### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [ ] Changes have been tested
- [ ] `Change-type` present on at least one commit
- [ ] `Signed-off-by` is present
- [ ] The PR complies with the [Open Embedded Commit Patch Message Guidelines](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)
<!-- optional: `Changelog-entry` present on at least one commit if you want to set the changelog entry manually-->

### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

resin-jenkins commented:
Can one of the admins verify this patch?

jakogut commented:
The first two commits unconditionally enable some BTRFS features in u-boot and the kernel, which is undesirable. These should only be enabled if the image uses BTRFS, but `RESIN_ROOT_FSTYPE` is a recipe variable with limited scope. To my understanding, this variable can't be accessed from other recipes. Furthermore, we may not want to couple feature selection to this variable.

@klutchell, @alexgg, Would it be reasonable to create a variable in, for example, local.conf that specifies filesystem support such as `BALENA_FILESYSTEMS="vfat ext4"`, or is there a better way to do this?

klutchell commented:
> The first two commits unconditionally enable some BTRFS features in u-boot and the kernel, which is undesirable. These should only be enabled if the image uses BTRFS, but `RESIN_ROOT_FSTYPE` is a recipe variable with limited scope. To my understanding, this variable can't be accessed from other recipes. Furthermore, we may not want to couple feature selection to this variable.
> 
> @klutchell, @alexgg, Would it be reasonable to create a variable in, for example, local.conf that specifies filesystem support such as `BALENA_FILESYSTEMS="vfat ext4"`, or is there a better way to do this?

I think a better way would be to add `btrfs` as an optional `DISTRO_FEATURES` entry that can set in `local.conf`, `distro.conf`, or `machine.conf`, etc and read by any recipe.

`${@bb.utils.contains('DISTRO_FEATURES', 'btrfs', 'if-enabled', 'if-not-enabled', d)}`

So if your goal is to conditionally enable/disable BTRFS features in other recipes without having access to `RESIN_ROOT_FSTYPE`, I think `DISTRO_FEATURES` is a clean way to do it.

jakogut commented:
@klutchell Perfect, that's the information I was looking for. Thanks!

alexgg commented:
@jakogut @klutchell We need a configuration variable that can specify different filesystem types. For example, devices with UBI filesystems might also be supported (this is something I am working on as a side project).
What about adding a `BALENA_FSTYPE` variable to the distro configuration that device types can then override in the device repositories? The variable will then be used to select the `RESIN_ROOTFS_TYPE` type, something like:
```
BALENA_FSTYPE = "ext4"
RESIN_ROOTFS_TYPE = "hostapp-${BALENA_FSTYPE}"
```
That is what I haven been doing with my UBIFS support branch.  It is then also used by the `docker-disk` recipe to create a UBIFS data disk image.

I know for the BTRFS/squashfs we have said the data disk those not need to change, but the changes should take that into account too.

jakogut commented:
@klutchell @alexgg I've updated this PR with conditional support for BTRFS dependent on the presence of `btrfs` in `DISTRO_FEATURES`.  I've also added the `BALENA_FSTYPE` conf that switches the root filesystem.

klutchell commented on meta-balena-common/recipes-containers/mkfs-hostapp-native/files/Dockerfile:
```diff
@@ -1,11 +1,15 @@
+# Fixes issue with Docker expecting iptables when the distro is using nftables
+RUN update-alternatives --set iptables /usr/sbin/iptables-legacy
 
-CMD [ "/usr/bin/create" ]
```
Maybe keep the default CMD in case it's part of another workflow and just change it to `"/usr/bin/create.ext4"`?

jakogut replied:
Fixed

alexgg commented:
@jakogut should this be a draft PR?

jakogut commented:
> @jakogut should this be a draft PR?

@alexgg I had removed the draft status, but after looking at it again, I do think the btrfs-tools issue needs figured out first.
